### DISCUSSION TITLE
Is there a way to add custom UI elements?

### DISCUSSION THREAD

**Discussion by ad-si** [Kong/insomnia#3467] [Help] [answered]
E.g. I'd like to have extra buttons to easily trigger a request. Or a textarea to easily post some text. Plugins don't seem to support anything like this.

**Comment by dimitropoulos** [comment]
I'm interested to hear more about what you have in mind.  Could you say a little more about what the use-case is?

**Reply by ad-si** [reply]
Once you've developed the whole API and defined all the endpoints in Insomnia, you've basically already created have a GUI for your service. E.g. submitting some data => Change the variables in the POST request and click send. That's of course a little unwieldy. But if you could have a few input fields and a button that would already be good enough for internal tooling.

**Reply by develohpanda** [reply]
That's pretty awesome, and I think I have a way forward for you!

- The [networking context](https://support.insomnia.rest/article/173-plugins#context.network) allows you to alter the request programatically ie. change the body or header or anything else
- Create a [workspace action](https://support.insomnia.rest/article/173-plugins#workspace-actions) or [folder action](https://support.insomnia.rest/article/173-plugins#folder-actions) so you have access to all of the requests within it ie. to send to the networking context
- The workspace action or folder action can open a [dialog](https://github.com/Kong/insomnia/blob/cdd279e09a16255cbe4681fc78db0690150e87e2/packages/insomnia-app/app/plugins/context/app.tsx#L33-L54) (which allows for custom HTML/GUI). You could even build that using React and the insomnia component library, through your plugin ie. the buttons and small configurations you need for your particular requests

Even if you choose not to create a modal with a GUI, you could have several workspace or folder actions which with one-click run your requests, or [prompt](https://github.com/Kong/insomnia/blob/cdd279e09a16255cbe4681fc78db0690150e87e2/packages/insomnia-app/app/plugins/context/app.tsx#L56-L85) for some info before running a request 

The [official docs](https://support.insomnia.rest/article/173-plugins) are a bit incomplete right now, but, it feels like your use case could be satisfied with the current state of the plugin system 🙌🏽

**Reply by develohpanda** [reply]
Alternatively, are you familiar with the prompt template tag already built in to Insomnia? We have an admin API configured with these prompts exactly as you describe 😄 they can be used anywhere - query params, headers, body, url, etc

![2021-06-15 09 39 59](https://user-images.githubusercontent.com/4312346/121963305-09649880-cdbe-11eb-9382-7e9148076010.gif)

**Reply by ad-si** [reply]
Awesome! Thanks for the elaboration. The prompt template tags look perfect 😎.

That's not how insomnia is marketed so far, but I think it would be the perfect starting point for a "internal tooling app" offering (aka competition to https://retool.com and similar offerings)

**Reply by develohpanda** [reply]
Awesome! That's a good take on it too, I know a few people have Insomnia set up in this way using template tags.

I'll add the template tag comment as a separate answer, then mark as the answer! Still getting used to discussions, lol

**Comment by develohpanda** [comment, marked as answer]
Are you familiar with the prompt template tag already built in to Insomnia? We have an admin API configured with these prompts exactly as you describe 😄 they can be used anywhere - query params, headers, body, url, etc

![2021-06-15 09 39 59](https://user-images.githubusercontent.com/4312346/121963305-09649880-cdbe-11eb-9382-7e9148076010.gif)
